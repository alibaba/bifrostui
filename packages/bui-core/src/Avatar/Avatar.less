.bui-avatar {
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 12px;
  background-color: var(--bui-color-bg-default);
  font-family: var(--bui-font-family);

  &-children {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  &-large {
    width: 60px;
    height: 60px;
  }

  &-medium {
    width: 42px;
    height: 42px;
  }

  &-small {
    width: 30px;
    height: 30px;
  }

  &-xsmall {
    width: 24px;
    height: 24px;
  }
}

.bui-avatar-rounded {
  border-radius: 50%;

  .bui-avatar-img {
    width: 100%;
    height: 100%;
    border-radius: 50%;
  }
}

.bui-avatar-square {
  border-radius: var(--bui-shape-radius-poster);

  .bui-avatar-img {
    border-radius: var(--bui-shape-radius-poster);
  }
}
