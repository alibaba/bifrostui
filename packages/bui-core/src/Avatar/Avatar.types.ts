import { OverrideProps } from '@bifrostui/types';
import React from 'react';
import { ImageProps } from '../Image/Image.types';

export type AvatarProps<
  D extends React.ElementType = 'div',
  P = {},
> = OverrideProps<
  {
    props: P & {
      /** 图像地址 */
      src?: ImageProps['src'];
      /**
       * 图像无法显示时的替代文本
       */
      alt?: ImageProps['alt'];
      /**
       * 加载时占位
       */
      placeholder?: ImageProps['placeholder'];
      /**
       * 懒加载
       */
      lazy?: ImageProps['lazy'];
      /**
       * 图片裁剪、缩放的模式
       */
      fit?: ImageProps['fit'];
      /**
       * 透传ImageProps
       */
      ImageProps?: ImageProps;
      /**
       * 头像形状
       */
      variant?: 'rounded' | 'square';
      /**
       * 头像大小
       */
      size?: 'large' | 'medium' | 'small' | 'xsmall';
      /**
       * 图像头像加载失败回调
       */
      onError?: ImageProps['onError'];
      /**
       * 图像头像加载回调
       */
      onLoad?: ImageProps['onLoad'];
    };
    defaultComponent: D;
  },
  D
>;
