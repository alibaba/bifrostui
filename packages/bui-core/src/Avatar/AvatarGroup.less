.bui-avatar-group {
  display: flex;
  flex-direction: row;
  font-family: var(--bui-font-family);

  &-avatar,
  &-extra {
    box-shadow: 0 0 0 2px var(--bui-color-bg-view);
    color: var(--bui-color-fg-default);
  }

  .bui-avatar-xsmall {
    &:not(:first-child) {
      margin-left: -5px;
    }
  }

  .bui-avatar-small {
    &:not(:first-child) {
      margin-left: -9px;
    }

    box-shadow: 0 0 0 1px var(--bui-color-bg-view);
  }

  .bui-avatar-large {
    &:not(:first-child) {
      margin-left: -21px;
    }
    box-shadow: 0 0 0 3px var(--bui-color-bg-view);
  }

  .bui-avatar-medium {
    &:not(:first-child) {
      margin-left: -13px;
    }
    box-shadow: 0 0 0 2px var(--bui-color-bg-view);
  }
}
