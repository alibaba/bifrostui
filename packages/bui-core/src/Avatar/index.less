@import './Avatar.less';
@import './AvatarGroup.less';
