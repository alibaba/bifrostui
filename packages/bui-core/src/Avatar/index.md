---
group:
  title: 基本输入
name: Avatar 头像
---

# Avatar 头像

用来展示用户头像或者事物，支持展示图片、文字、图标等。
`Avatar` 基于 `Image`组件进行封装，[Image](http://bui.taopiaopiao.com/cores/image) 常用的属性，Avatar 同样支持。

## 代码演示

### 基础用法

头像组件通过`src`属性传递图片地址来创建图像头像。

```tsx
import { Avatar, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack>
      <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
    </Stack>
  );
};
```

### 字符串头像

除了图片头像外，您还可以通过`children`创建简单字符型头像。

```tsx
import { Avatar, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="10px">
      <Avatar>UC</Avatar>
      <Avatar>UFCEAEF</Avatar>
    </Stack>
  );
};
```

### 自定义头像

通过`children`还可以创建图标等自定义头像。

```tsx
import React from 'react';
import { Avatar, Stack } from '@bifrostui/react';
import { VipFilledIcon } from '@bifrostui/icons';

export default () => {
  return (
    <Stack direction="row" spacing="10px">
      <Avatar>
        <VipFilledIcon />
      </Avatar>
    </Stack>
  );
};
```

### 头像形状

Avatar 组件提供两种头像模式：圆形`rounded` 、方形`square`， 您可以使用`variant`设置对应的模式。
variant 默认值为：`rounded`。

```tsx
import { Avatar, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="10px">
      <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      <Avatar
        variant="square"
        src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png"
      />
    </Stack>
  );
};
```

### 头像尺寸

`Avatar`提供`xsmall`、 `small`、 `medium`、 `large` 等 4 种尺寸，您可以通过`size`设置对应尺寸。
size 默认尺寸为： `medium`。

```tsx
import { Avatar, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="row" spacing="10px">
      <Avatar
        src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png"
        size="xsmall"
      />
      <Avatar
        src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png"
        size="small"
      />
      <Avatar
        src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png"
        size="medium"
      />
      <Avatar
        src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png"
        size="large"
      />
    </Stack>
  );
};
```

## 头像组合

除了提供单个头像，我们还提供组合头像，通过`AvatarGroup`、`Avatar`组合使用。

```tsx
import { Avatar, AvatarGroup, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack>
      <AvatarGroup>
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
    </Stack>
  );
};
```

### 头像形状

AvatarGroup 组件提供两种头像模式：圆形`rounded` 、方形`square`， 您可以使用`variant`设置对应的模式。
variant 默认值为：`rounded`。

```tsx
import { Avatar, AvatarGroup, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="column" spacing="10px">
      <AvatarGroup>
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
      <AvatarGroup variant="square">
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
    </Stack>
  );
};
```

### 控制头像显示数目

AvatarGroup 组件提供`maxCount`属性来限制显示的头像数量，并自动计算未显示的头像数量。

```tsx
import { Avatar, AvatarGroup, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack>
      <AvatarGroup maxCount="2">
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
    </Stack>
  );
};
```

### 头像组合朝向

AvatarGroup 组件提供`orientation`属性来控制头像朝向。
orientation 默认为：`left`。

```tsx
import { Avatar, AvatarGroup, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack direction="column" spacing="10px">
      <AvatarGroup>
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
      <AvatarGroup orientation="left">
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
    </Stack>
  );
};
```

### 头像间距

AvatarGroup 组件提供`spacing`属性来控制头像间距。

```tsx
import { Avatar, AvatarGroup, Stack } from '@bifrostui/react';
import React from 'react';

export default () => {
  return (
    <Stack>
      <AvatarGroup maxCount="2" spacing="10px">
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
        <Avatar src="https://gw.alicdn.com/tfs/TB1Q2zRXuL2gK0jSZFmXXc7iXXa-120-120.png" />
      </AvatarGroup>
    </Stack>
  );
};
```

### API

##### AvatarProps

| 属性        | 说明                     | 类型                                                 | 默认值    |
| ----------- | ------------------------ | ---------------------------------------------------- | --------- |
| src         | 图像地址                 | string                                               | -         |
| alt         | 图像无法显示时的替代文本 | React.ReactNode                                      | -         |
| lazy        | 懒加载                   | boolean                                              | false     |
| placeholder | 加载时占位               | React.ReactNode ｜ boolean                           | false     |
| ImageProps  | 透传 Image 属性          | [ImageProps](http://bui.taopiaopiao.com/cores/image) | -         |
| variant     | 头像形状                 | `rounded`｜`square`                                  | `rounded` |
| size        | 头像大小                 | `large`｜`medium`｜`small` ｜`xsmall`                | `medium`  |
| onError     | 图像头像加载失败回调     | ReactEventHandler                                    | -         |
| onLoad      | 图像头像加载回调         | ReactEventHandler                                    | -         |

##### AvatarGroupProps

| 属性        | 说明                 | 类型                                  | 默认值    |
| ----------- | -------------------- | ------------------------------------- | --------- |
| maxCount    | 最多可显示的头像个数 | string                                | -         |
| variant     | 头像形状             | `rounded` ｜`square`                  | `rounded` |
| size        | 头像大小             | `large`｜`medium`｜`small` ｜`xsmall` | `medium`  |
| spacing     | 头像间距             | string                                | -         |
| orientation | 头像朝向             | `left`｜`right`                       | `right`   |
