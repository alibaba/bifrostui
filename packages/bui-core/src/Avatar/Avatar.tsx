import clsx from 'clsx';
import React from 'react';
import Image from '../Image';
import { AvatarProps } from './Avatar.types';
import './Avatar.less';

const prefixCls = 'bui-avatar';

const Avatar = React.forwardRef<HTMLDivElement, AvatarProps>((props, ref) => {
  const {
    className,
    onClick,
    onError,
    onLoad,
    children,
    src,
    lazy,
    variant,
    size,
    alt,
    fit,
    placeholder,
    ImageProps,
    ...others
  } = props;

  const AvatarImg = src ? (
    <Image
      src={src}
      lazy={lazy}
      alt={alt}
      fit={fit}
      placeholder={placeholder}
      onError={onError}
      onLoad={onLoad}
      {...ImageProps}
      className={clsx(ImageProps?.className, `${prefixCls}-img`)}
    >
      {children}
    </Image>
  ) : null;

  // flex布局容器直接设置超出省略不起效
  const createElement = (element) =>
    element ? <div className={`${prefixCls}-children`}>{element}</div> : null;

  return (
    <div
      className={clsx(
        prefixCls,
        className,
        `${prefixCls}-${size}`,
        `${prefixCls}-${variant}`,
      )}
      onClick={onClick}
      ref={ref}
      {...others}
    >
      {createElement(children) || AvatarImg}
    </div>
  );
});

Avatar.displayName = 'BuiAvatar';
Avatar.defaultProps = {
  variant: 'rounded',
  size: 'medium',
  fit: 'cover',
};

export default Avatar;
