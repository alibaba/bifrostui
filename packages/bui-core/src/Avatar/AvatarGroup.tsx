import clsx from 'clsx';
import React, { cloneElement, ReactElement } from 'react';
import Avatar from './Avatar';
import './AvatarGroup.less';
import { AvatarGroupProps } from './AvatarGroup.types';

const prefixCls = 'bui-avatar-group';
const AvatarGroup = React.forwardRef<HTMLDivElement, AvatarGroupProps>(
  (props, ref) => {
    const {
      className,
      maxCount,
      variant,
      spacing,
      size,
      orientation,
      children,
      onClick,
      ...others
    } = props;
    const avatarChildrenCount = React.Children.count(children);
    const extraAvatarNum = avatarChildrenCount - Number(maxCount);

    const AvatarGroupItem = children
      ? React.Children.map(children, (child: ReactElement, index) => {
          if (maxCount && index > Number(maxCount)) return null;
          const zIndex =
            orientation === 'right' ? -index + avatarChildrenCount : index;

          if (maxCount && index === Number(maxCount)) {
            return (
              <Avatar
                variant={variant}
                size={size}
                className={`${prefixCls}-extra`}
                style={{ marginLeft: `-${spacing}`, zIndex }}
                onClick={onClick}
              >{`+${extraAvatarNum}`}</Avatar>
            );
          }

          const childStyle = {
            className: clsx(`${prefixCls}-avatar`, child?.props?.className),
            style: {
              ...child?.props?.style,
              marginLeft: `-${spacing}`,
              zIndex,
            },
            size,
            variant,
          };

          return React.isValidElement(child)
            ? cloneElement(child, { ...childStyle })
            : child;
        })
      : '';

    return AvatarGroupItem ? (
      <div className={clsx(prefixCls, className)} ref={ref} {...others}>
        {AvatarGroupItem}
      </div>
    ) : null;
  },
);

AvatarGroup.displayName = 'BuiAvatarGroup';
AvatarGroup.defaultProps = {
  size: 'medium',
  orientation: 'right',
};

export default AvatarGroup;
