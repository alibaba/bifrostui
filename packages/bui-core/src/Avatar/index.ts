import AvatarGroup from './AvatarGroup';

export * from './Avatar.types';
export * from './AvatarGroup.types';
export { default } from './Avatar';
export { default as Avatar } from './Avatar';
export { AvatarGroup };
