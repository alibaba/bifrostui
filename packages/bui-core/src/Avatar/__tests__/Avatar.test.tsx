import React from 'react';
import { isConformant, render, screen, userEvent } from 'testing';
import Avatar from '../Avatar';

const classPrefix = 'bui-avatar';

describe('Avatar', () => {
  isConformant({
    Component: Avatar,
    displayName: 'BuiAvatar',
    className: 'bui-avatar',
  });
  describe('prop: src', () => {
    it('should render children correctly, not src', () => {
      const component = (
        <Avatar src="https://gw.alicdn.com/imgextra/i3/O1CN01tp2YUs1WFD8lboMDu_!!6000000002758-2-tps-126-126.png">
          hello world
        </Avatar>
      );
      const { container } = render(component);

      expect(screen.getByText('hello world')).toBeTruthy();
      expect(screen.queryByRole('img')).toBe(null);
      expect(container).toMatchSnapshot();
    });

    it('should render containing an img', () => {
      const component = (
        <Avatar src="https://gw.alicdn.com/imgextra/i3/O1CN01tp2YUs1WFD8lboMDu_!!6000000002758-2-tps-126-126.png" />
      );
      const { container } = render(component);

      expect(screen.getByRole('img')).toHaveAttribute(
        'src',
        'https://gw.alicdn.com/imgextra/i3/O1CN01tp2YUs1WFD8lboMDu_!!6000000002758-2-tps-126-126.png',
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('prop: size', () => {
    it.each(['xsmall', 'small', 'medium', 'large'] as const)(
      'should render size correct',
      (size) => {
        render(<Avatar size={size} data-testid="avatar" />);
        expect(screen.getByTestId('avatar')).toHaveClass(
          `${classPrefix}-${size}`,
        );
      },
    );
  });

  describe('prop: variant', () => {
    it('should render rounded default correctly', () => {
      render(<Avatar data-testid="avatar" />);
      expect(screen.getByTestId('avatar')).toHaveClass(
        `${classPrefix}-rounded`,
      );
    });
    it('should render square correctly', () => {
      render(<Avatar variant="square" data-testid="avatar" />);
      expect(screen.getByTestId('avatar')).toHaveClass(`${classPrefix}-square`);
    });
  });

  describe('prop: ImageProps', () => {
    render(
      <Avatar src="https://not.exists.image/" ImageProps={{ fit: 'fill' }} />,
    );
    expect(screen.getByRole('img').style.objectFit === 'fill');
  });
  // it('has placeholder', async () => {
  //   render(<Avatar src="https://not.exists.image/" placeholder="加载失败" />);
  //   expect(screen.getByText('加载失败')).toBeTruthy();
  // });

  it.each(['contain', 'cover', 'fill', 'none', 'scale-down'])(
    'renders with different fit modes',
    async (fit: any) => {
      render(
        <Avatar
          src="https://gw.alicdn.com/imgextra/i3/O1CN01tp2YUs1WFD8lboMDu_!!6000000002758-2-tps-126-126.png"
          fit={fit}
        />,
      );
      expect(screen.getByRole('img').style.objectFit === fit);
    },
  );

  describe('events', () => {
    it('responses click events', () => {
      const onClick = jest.fn();
      render(<Avatar onClick={onClick}>hello world</Avatar>);
      userEvent.click(screen.getByText('hello world'));
      expect(onClick).toHaveBeenCalled();
    });
  });
});
