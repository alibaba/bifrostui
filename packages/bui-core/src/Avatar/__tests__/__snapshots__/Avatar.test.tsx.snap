// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar prop: src should render children correctly, not src 1`] = `
<div>
  <div
    class="bui-avatar bui-avatar-medium bui-avatar-rounded"
  >
    <div
      class="bui-avatar-children"
    >
      hello world
    </div>
  </div>
</div>
`;

exports[`Avatar prop: src should render containing an img 1`] = `
<div>
  <div
    class="bui-avatar bui-avatar-medium bui-avatar-rounded"
  >
    <div
      class="bui-image bui-avatar-img"
    >
      <img
        class="bui-image-img bui-image-loaded"
        loading="eager"
        src="https://gw.alicdn.com/imgextra/i3/O1CN01tp2YUs1WFD8lboMDu_!!6000000002758-2-tps-126-126.png"
        style="object-fit: cover;"
      />
    </div>
  </div>
</div>
`;
